### ISSUE TITLE
Moving the webview brigde to React-Native Core

### ISSUE THREAD

**Issue by alinz** [alinz/react-native-webview-bridge#109]
Hello guys,

I have decided that, it is time to move this project into core. one of the problem with maintaining this project is not be able to extend the native class and as a result I have to copy and paste the entire source code of WebView implementation of react-native. This makes it very difficult. So what I'm asking is goto this [link](https://productpains.com/post/react-native/bridge-for-webview-and-react-native) and vote that one. If I get the confirmation from core-team I will make a PR request to WebView and I will merge this project to core.

**Comment by stereodenis** [comment]
@alinz it is good idea! But could you check PRs?)

**Comment by goldenice** [comment]
This should be a core functionality for sure, but perhaps a more high level approach (e.g. executing arbitrary JS in the webview) would be useful. I'm not sure how that would work on iOS, but I do know that would be easy to implement on Android. Writing your own bridging implementation (at least RN to webview) would be trivial if that kind of communication is possible

**Comment by alinz** [comment]
@goldenice RN's WebView has already implemented run arbitrary JS. what I want to add is the reverse one. WebVIew send a message back to RN. If you look into the implementation of WebView, you will see most of the important classes are inlined and because of that you can't extend it. That's what I want to do in core.

**Comment by stereodenis** [comment]
@alinz https://github.com/alinz/react-native-webview-bridge/pull/90

**Comment by fungilation** [comment]
You have my vote. But approve pending PRs pretty please

**Comment by Naoto-Ida** [comment]
+1

**Comment by Hamiltontx** [comment]
+1

**Comment by bbay** [comment]
+1

**Comment by tslater** [comment]
Is there a PR for the react-native core?

**Comment by alinz** [comment]
working on it. should be send to them by couple of days.

**Comment by huangciyin** [comment]
+1

**Comment by alinz** [comment]
Since I'm still behind and react-native 30 is just around the corner, I will delay this and release the new version of app this weekend. Extremely busy with other stuff but I will find a time this weekend and make it available for you guys.

**Comment by CaoLiang1991** [comment]
+1

**Comment by starlight36** [comment]
+1

**Comment by stereodenis** [comment]
@alinz we're waiting))

**Comment by imlooke** [comment]
+1

**Comment by benlinton** [comment]
+1

**Comment by derdav3** [comment]
+1

**Comment by HectorFL** [comment]
+1

**Comment by dfejgelis** [comment]
RN 0.30 is here :)

**Comment by fungilation** [comment]
Is it including the bridge integrated?

**Comment by derdav3** [comment]
doesnt seem like it...? -> [release notes](https://github.com/facebook/react-native/releases/tag/v0.30.0)

**Comment by ivanakimov** [comment]
+1

**Comment by jzimmek** [comment]
+1

**Comment by cbcye** [comment]
+1 for this feature

**Comment by kinergy** [comment]
+1

**Comment by Roman-K93** [comment]
+1

**Comment by avishayil** [comment]
+1

**Comment by Fasani** [comment]
+1

**Comment by anhuin69** [comment]
+1

**Comment by daikini** [comment]
+1

**Comment by rbmedia** [comment]
+1

**Comment by musicode** [comment]
does it finished?

**Comment by saulshanabrook** [comment]
@musicode The core team haven't weighed in on whether they want this in React and there has been no public progress on integrating the code.

**Comment by juankiz** [comment]
Hey @alinz have you found time to work on the pull request? If you need more hands let us know, I'd be glad to help if needed :)

**Comment by avishayil** [comment]
@alinz Are you following here mate?

**Comment by alidawud** [comment]
Hi Mr. @alinz any good news?

**Comment by MacKentoch** [comment]
Some news here [#141](https://github.com/alinz/react-native-webview-bridge/issues/141)

**Comment by stereodenis** [comment]
https://github.com/facebook/react-native/pull/9762

**Comment by fungilation** [comment]
Hmm https://github.com/facebook/react-native/pull/9762#issuecomment-246062195

**Comment by Etheryte** [comment]
I wonder what a reasonable estimate would be on the above link getting merged? I need the functionality, but it would seem a waste to merge the different fixing branches from here only to have the functionality available in the main package shortly after.

**Comment by fungilation** [comment]
RN releases every 2 weeks. So in 0.34 or 0.35?

**Comment by adamkdean** [comment]
Still not release. Last activity 7 days ago. 😭

**Comment by alinz** [comment]
There is an issue in this PR https://github.com/facebook/react-native/pull/6478 which holds me back to send a PR to core. Core team still debating about the right API and I'm waiting for that. In a meantime, I was working on RPC implementation which significantly simplify the process of calling methods on both WebView and React-Native. Checkout out the v2 branch.

in terms of release, I won't release any version until WebView changes in react-native side.

**Comment by tioback** [comment]
@alinz in the mean time, can we use branch v2?

**Comment by alinz** [comment]
@tioback yes, you can, v2 branch is being used in production at the moment and we are using RPC implementation on top of that. Please refer to the rpc example on that branch.

**Comment by adamipc** [comment]
@alinz Is v2 ready on Android? I notice the README for v2 is a bit sparse, leaving out the instructions from master that I assume are still required. But it doesn't say how to specifically add the javascript asset files on both Xcode and Android, If it DOES work on Android, could you at least write up a little how to on where to put the js files so they can be loaded? Thanks!

**Comment by Etheryte** [comment]
@adamipc If you look at the example project, you can see that you need to add the two script files directly into your main project in Xcode. Haven't tested it in Android.

Agreed, that better documentation would help, but all in due time.

**Comment by adamipc** [comment]
@Etheryte Thanks, you're right, the example project does show where to add stuff and I should be able to use that to figure it out (it does appear to have Android as well so I'll test for myself whether it works)

**Comment by GantMan** [comment]
@alinz - Awesome work, we all appreciate you adding this to core.

**Comment by seakingii** [comment]
+1

**Comment by quadsurf** [comment]
I see that React Native has the 'injectedJavaScript' property/attribute included in WebView:

<WebView injectedJavaScript={} />

Does this deliver the same capability/objective as  @alinz 's  webview bridge? If not, how do they differ? (I'm still a coding noob)

**Comment by adamipc** [comment]
@quadsurf The injectedJavaScript lets you run arbitrary javascript within the WebView but doesn't give you any mechanism for communicating between the WebView and your React Native code which is what the WebView Bridge adds.

**Comment by quadsurf** [comment]
thank you @adamipc ... since alinz added this to core via a recent PR, would updating to RN @latest get me all the functionality from this repo? Or would I still need to install this repo if I need that RN-to-WebView communication.

**Comment by stereodenis** [comment]
https://github.com/facebook/react-native/commit/abb8ea3aea686e2cd881a61fbc66d137857ef422

**Comment by tioback** [comment]
@stereodenis, care to explain what you aimed to achieve by simply posting a link?

**Comment by jacobp100** [comment]
It’s been added in 0.37.

**Comment by mlostekk** [comment]
wonderful! thanks to all for their great work!

**Comment by fungilation** [comment]
👍🏻

**Comment by Etheryte** [comment]
Please be considerate when replying to issues with many participants.  
If you can, use the existing reaction functionality instead of adding a new comment, otherwise 40 people will get an email with just a thumbs-up emoji in their inbox.

**Comment by 3449790992** [comment]
Javascript alert not working in android react-native webview，can you update it for support js alert?

http://stackoverflow.com/questions/41028001/javascript-alert-not-working-in-android-react-native-webview

**Comment by danielearwicker** [comment]
So do I need this package or is it already included in react-native? Above there are several messages indicating the project was added to core and then other messages saying it isn't yet.

**Comment by Etheryte** [comment]
@danielearwicker Did you try reading [the docs](https://facebook.github.io/react-native/docs/webview.html)?

**Comment by danielearwicker** [comment]
@Etheryte yes! I'm playing with WebView right now. I'm wondering if this project has anything extra.

**Comment by fungilation** [comment]
I'm still using this as RN's own is still stuck in the ivory tower: https://github.com/facebook/react-native/pull/10941#issuecomment-278807506

Unfortunately, as this package is getting orphaned with no merging of critical PRs.

**Comment by sun2rise** [comment]
@alinz thanks for your work and kudos for the integration!
do you know if `allowFileAccessFromFileURLs` and `allowUniversalAccessFromFileURLs` features will still be available? they add a great value to the package i think but i cannot use those properties with the RN official package...
or maybe you know how to workaround this problem once and for all?

**Comment by naltimari** [comment]
it seems the 'v2' branch was NOT added to React Native's mainline code. The 'rpc' feature is very interesting and exactly what I need.

**Comment by tioback** [comment]
@naltimari, almost that. They didn't merge the v2 branch "as is", however the feature is there.
You can still do the to and fro communication with the WebView.

**Comment by naltimari** [comment]
@tioback the rpc feature allows me to return a promise to the 'caller' and then resolve it from inside the react app (passing the result of the function call); it's almost like a sync function call, which is what I want

**Comment by SiDevesh** [comment]
@tioback @naltimari Could you direct me to where I can get to know more about this rpc feature? Is this something built in react native now ?

**Comment by naltimari** [comment]
@SiDevesh check the v2 branch of the project. The rpc allows you to offer an API to the webview so it can call methods from your app (that's what I'm working on). Unfortunately the v2 branch is still not compatible with RN40+, so I had to resort to use vext's fork of the master branch, and then implement an rpc-like feature in the injectScript

**Comment by SiDevesh** [comment]
@naltimari  okay thanks!!

**Comment by aqnaruto** [comment]
thankyou great work, does this support cache?

**Comment by vvavepacket** [comment]
so whats the status for this on react native 0.45 and above? was this integrated in core? i cant see on documentation that this supports android.

**Comment by yanhaijing** [comment]
+1

**Comment by jariwalabhavesh** [comment]
+1

**Comment by markeins** [comment]
Hi, so any news when this will be merged into core?

**Comment by hughlang** [comment]
+1

**Comment by neiker** [comment]
+1

**Comment by NaveenDK** [comment]
@alinz , just wondering if we could use this with expo react native? Thanks

**Comment by Barylskyigb** [comment]
The core has this feature now

**Comment by fungilation** [comment]
No it doesn't. Core has a f'ed up version with .postMessage that gets overrided by websites in the wild, rendering it useless.

**Comment by sunny635533** [comment]
+1

**Comment by bhargavb07** [comment]
+1

**Comment by alihesari** [comment]
+1

**Comment by fungilation** [comment]
For sure. The whole RN community owes you @jamonholmgren a great deal in your spearheading effort there, consolidating contributions into a community developed WebView that doesn't suck for all platforms.

I'm not a native dev however so I can't help with the native side. What's the maturity of WKWebView in https://github.com/react-native-community/react-native-webview? Is it time yet to discuss a more robust `postMessage`, `onMessage` so ".postMessage (in WebView) wouldn't get overridden by websites in the wild"? This 2 way communication between RN and javascript within WebViews is what I need in my app, and I'll help with what I can in getting that done so I/we could migrate from react-native-webview-bridge.

https://github.com/CRAlpha/react-native-wkwebview seems to have a pretty robust API that's iOS for this RN<>WebView communication. But from what I remember when I tried using it, it had issues and just didn't work right with message passing (I don't remember exactly how, it was a while ago too). But if it works better now, it can be the basis for expanding it for Android's side of WebView that would share the same RN<>WebView bridge api.

**Comment by fungilation** [comment]
I started an issue at https://github.com/react-native-community/react-native-webview/issues/66 to discuss this further.

**Comment by kirill578** [comment]
I've written a wrapper the simplifies calling methods between react-native and the DOM.
check it out:
https://github.com/kirill578/react-native-webview-bridge-seamless

**Comment by alinz** [comment]
@kirill578 why your project is better than this? Are you supporting all features? if then can you send us PR to make it better?

**Comment by kirill578** [comment]
The requirement for a project I am working on is to be able to reuse existing code written in `react native` within the `webview`. For example, an API call from the webview requires a token, while there is a function in the react project called `generateToken()` that will take care of refreshing and authenticating the user.

I was considering to use this library in the beginning, but as the requirements to reuse the code grew, the switch statement inside `onBridgeMessage` grew as well. the issue became bigger as I had to handle promise rejections inside `onBridgeMessage`. if the `generateToken` failed for some reason I had to serialize the exception and resend it via `sendToBridge`

at the end of the day it looked to me that there is bunch of boiler plate code, so the idea behind `react-native-webview-bridge-seamless` is to facilitate inter process function call, between `react-native` and the `webview`.

adding new code that can be reused now, is a single liner. and the library handles all the serialization of successful or unsuccessful function invocation, which will also support rejecting an invocation with a timeout.

Last but not least, it's built as a wrapper for `react-native-webview` maintained by `react-native-community` which has support for more features like photo picker, photo taking from the webview, which I believe is something we will have to use at some point.

Ideally it could become a PR  for `react-native-webview`, to avoid multiple projects that do the same thing.
